src/virtual.ts

```diff
@@ -1,78 +1,95 @@
-import { directive, NodePart } from 'lit-html';
+import { directive, DirectiveParameters, ChildPart, PartInfo } from 'lit/directive.js';
+import { noChange } from 'lit';
+import { AsyncDirective } from 'lit/async-directive.js';
 import { GenericRenderer } from './core';
 import { BaseScheduler } from './scheduler';
 
 const includes = Array.prototype.includes;
 
-interface Renderer extends GenericRenderer<NodePart> {
-  (this: NodePart, ...args: unknown[]): unknown | void;
+interface Renderer extends GenericRenderer<ChildPart> {
+  (this: ChildPart, ...args: unknown[]): unknown | void;
 }
 
-function makeVirtual() {
-  const partToScheduler: WeakMap<NodePart, Scheduler> = new WeakMap();
-  const schedulerToPart: WeakMap<Scheduler, NodePart> = new WeakMap();
+const partToScheduler: WeakMap<ChildPart, Scheduler> = new WeakMap();
+const schedulerToPart: WeakMap<Scheduler, ChildPart> = new WeakMap();
 
-  class Scheduler extends BaseScheduler<object, NodePart, Renderer, NodePart> {
-    args!: unknown[];
+class Scheduler extends BaseScheduler<object, ChildPart, Renderer, ChildPart> {
+  args!: unknown[];
+  setValue: Function;
 
-    constructor(renderer: Renderer, part: NodePart) {
-      super(renderer, part);
-      this.state.virtual = true;
-    }
+  constructor(renderer: Renderer, part: ChildPart, setValue: Function) {
+    super(renderer, part);
+    this.state.virtual = true;
+    this.setValue = setValue;
+  }
 
-    render(): unknown {
-      return this.state.run(() => this.renderer.apply(this.host, this.args));
-    }
+  render(): unknown {
+    return this.state.run(() => this.renderer.apply(this.host, this.args));
+  }
 
-    commit(result: unknown): void {
-      this.host.setValue(result);
-      this.host.commit();
-    }
+  commit(result: unknown): void {
+    this.setValue(result);
+  }
 
-    teardown(): void {
-      super.teardown();
-      let part = schedulerToPart.get(this);
-      partToScheduler.delete(part!);
-    }
+  teardown(): void {
+    super.teardown();
+    let part = schedulerToPart.get(this);
+    partToScheduler.delete(part!);
   }
+}
+
+function makeVirtual() : any {
 
   function virtual(renderer: Renderer) {
-    function factory(...args: unknown[]) {
-      return (part: NodePart): void => {
-        let cont = partToScheduler.get(part);
-        if(!cont) {
```

Comment:
Checking to be sure `cont.renderer` matches `renderer` before reusing it fixes issue #337 , e.g.

```js
if (!cont || cont.renderer !== renderer) {
```

Since this is an existing bug, happy to submit a PR against the current release if you'd prefer. Just figured it might be easy to fix here while we're upgrading anyway.

Reply:
Put that check in and added a test to confirm it works.

Reply:
Awesome, thanks!